README.TXT
This is version 3 of the TEI-EMACS installation: a more or less
complete SGML/XML authoring system, which combines GNU-Emacs
with PSGML and a host of other relevant emacs customizations for
writing and validating SGML or XML documents. Most XML-emacs subsystems
have their own help system or documentation.

The systems included here are all distributed under the terms of the 
GNU GENERAL PUBLIC LICENSE (see the file info/COPYING). Individual
components also carry their own README files, which should be read for
further information.

This package includes SGML and XML DTDs and stylesheets for the latest
available versions of the TEI DTDs (TEI P4, published May 2002) and
also of other popular DTDs (Docbook, XHTML). Select the DTD you want
from the DTD menu.

This package was put together by 

 Lou Burnard (lou.burnard@oucs.ox.ac.uk) 
 Sebastian Rahtz (sebastian.rahtz@oucs.ox.ac.uk)
 Christian Wittern (wittern@kanji.zinbun.kyoto-u.ac.jp)

and is available from the TEI web site at
http://www.tei-c.org/Software/tei-emacs.

Please send any comments or suggestions for improvement to tei@oucs.ox.ac.uk

Last updated: 20 May 2003




debian/changelog
tei-emacs (3.11) karmic; urgency=low

  * remove speedbar, and P4

 -- Sebastian Rahtz <sebastian.rahtz@it.ox.ac.uk>  Mon, 11 Mar 2013 05:07:29 -0700

tei-emacs (3.10) karmic; urgency=low

  * remove nxml-mode dependency

 -- Sebastian Rahtz <sebastian.rahtz@it.ox.ac.uk>  Mon, 04 Feb 2013 05:07:29 -0700

tei-emacs (3.9) karmic; urgency=low

  * ref to xsl packages

 -- Sebastian Rahtz <sebastian.rahtz@it.ox.ac.uk>  Sun, 03 Feb 2013 05:07:29 -0700

tei-emacs (3.7) karmic; urgency=low

  * wrong version

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Sun, 25 Apr 2010 05:07:29 -0700

tei-emacs (emacs23) karmic; urgency=low

  * emacs23 

 -- Sebastian Rahtz <rahtz@ubuntu>  Sun, 25 Apr 2010 04:26:12 -0700

tei-emacs (3.6) jaunty; urgency=low

  * fixing schemas.xml

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Fri, 20 Feb 2009 18:59:13 +0000

tei-emacs (3.5) hardy; urgency=low

  * force emacs22

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Sat, 26 Apr 2008 13:13:33 +0100

tei-emacs (3.4) gutsy; urgency=low

  * drop unicode special stuff

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Tue, 27 Nov 2007 09:59:21 +0000

tei-emacs (3.3.8) feisty; urgency=low

  * emacs22

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Tue, 18 Sep 2007 13:36:20 +0100

tei-emacs (3.3.7) unstable; urgency=low

  * attempt to dependencies again

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Tue, 16 Jan 2007 14:49:32 +0000

tei-emacs (3.3.6) unstable; urgency=low

  * upstream add back TEI Lite

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Thu, 29 Jun 2006 23:39:45 +0100

tei-emacs (3.3.5) unstable; urgency=low

  * cleaning

 -- Sebastian Rahtz <rahtz@localhost.localdomain>  Sun, 22 Jan 2006 12:05:04 +0000

tei-emacs (3.3.4) unstable; urgency=low

  *  small fixes for Windows compatibiliyuy

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Tue, 29 Nov 2005 13:06:38 +0000

tei-emacs (3.3.3) unstable; urgency=low

  * location of skeleton files

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Fri,  8 Jul 2005 10:04:28 +0100

tei-emacs (3.3.2-3) unstable; urgency=low

  * change list of exemplars

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Thu, 30 Jun 2005 21:14:12 +0100

tei-emacs (3.3.2-2) unstable; urgency=low

  * correct path for ODD skeleton

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Mon, 27 Jun 2005 21:00:32 +0100

tei-emacs (3.3.2-1) unstable; urgency=low

  * add dependency on tei-lite and tei-p5-skeletons; redo the way skeletons work

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Fri, 24 Jun 2005 13:21:37 +0100

tei-emacs (3.3.1-2) unstable; urgency=low

  * stop xslt-process by default as it screws up on stable

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Fri, 24 Jun 2005 13:00:34 +0100

tei-emacs (3.3.1-1) unstable; urgency=low

  * New upstream release. changed location of skeletons

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Tue, 19 Apr 2005 03:36:38 +0100

tei-emacs (3.2.1-1) unstable; urgency=low

  * New upstream release

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Fri, 21 Jan 2005 17:02:03 +0000

tei-emacs (3.0.4-1) unstable; urgency=low

  * add dependency on tei-doc

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Wed, 27 Oct 2004 12:58:13 +0100

tei-emacs (3.0.3-1) unstable; urgency=low

  * New upstream release

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Mon, 11 Oct 2004 21:06:34 +0100

tei-emacs (3.0.2-1) unstable; urgency=low

  * New upstream release

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Thu, 23 Sep 2004 19:48:12 +0100

tei-emacs (3.0.1-1) unstable; urgency=low

  * upstream change, mistake in naming skeletons

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Thu, 23 Sep 2004 17:08:09 +0100

tei-emacs (3.0.0-3) unstable; urgency=low

  * add dependency on nxml-mode-tei, and conflict with nxml-mode

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Thu, 23 Sep 2004 12:33:05 +0100

tei-emacs (3.0.0-2) unstable; urgency=low

  * added dependency on tei-doc

 -- Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>  Thu, 23 Sep 2004 11:27:33 +0100

tei-emacs (3.0.0-1) unstable; urgency=low

  * Initial Release.

 -- Sebastian Rahtz <sebastian.rahtz@computing-services.oxford.ac.uk>  Sat, 28 Aug 2004 17:12:41 +0100



tei-emacs/xmlrun.el
;;;; xmlrun.el -- run various programs on an XML file
;; Sebastian Rahtz
;; copied from xslide-process.el by Tony Graham
;; Revised: $Date: 2004/09/23 13:23:17 $
(defcustom xmlrun-offer-save t
  "*If non-nil, ask about saving modified buffers before \\[xmlrun-run] is run."
  :type '(choice (const :tag "Yes" t) (const :tag "No" nil))
  :group 'xmlrun-run)

(defcustom xmlrun-command
  (list
   ;; xsltproc
   "xsltproc -o %o %s %i"
   ;; simplest case of embedded stylesheet PI
   "saxon -a %i"
   ;; Instant Saxon
   "saxon -o %o %i %s"
   ;; Instant Saxon using xml-stylesheet PI and output
   "saxon -o %o -a %i"
   ;; Saxon
   "java com.icl.saxon.StyleSheet -o %o %i %s"
   ;; Saxon using xml-stylesheet PI
   "java com.icl.saxon.StyleSheet -a -o %o %i"
   ;; validate using RXP
   "rxp -V -s %i"
   ;; validate using xmllint
   "xmllint --noout --valid %i"
)
  "*The shell command to process an XSL document.

This is a `format' control string that by default should contain three
`%s' conversion specifications: the first will be replaced by the
value of xmlrun-input-file \(or the empty string, if nil\); the
second will be replaced by xmlrun-stylesheet-file \(or the empty
string, if nil\); the third will be replaced by
xmlrun-output-file \(or the empty string, if nil\).

If `xmlrun-files' is non-nil, the format string should contain
one `%s' conversion specification for each element of its result.

If xmlrun-command is a list, then every element should be a
string.  The strings will be tried in order and %-sequences in the
string will be replaced according to the list below, if the string contains
%-sequences with no replacement value the next string will be tried.

%b means the visited file of the current buffer
%i means the value of xmlrun-input-file
%s means the value of xmlrun-stylesheet-file
%o means the value of xmlrun-output-file
"
  :type '(repeat :tag "Commands" string)
  :group 'xmlrun)

(defvar xmlrun-files nil
  "If non-nil, a function of no arguments that returns a list of file names.
These file names will serve as the arguments to the `xmlrun-command'
format control string instead of the defaults.")

(defcustom xmlrun-error-regexps
  '(("file:\\([^:]+\\):\\([0-9]+\\):\\([0-9]+\\):" 1 2 3)
    ("file:/\\(\\([A-Za-z]:\\)?[^:]+\\):\\([0-9]+\\):\\(\\([0-9]+\\):\\)?" 1 3 5)
    ("on line \\([0-9]+\\) of file:\\([^:]+\\):\\([0-9]+\\)" 3 1 2)
    ("on line \\([0-9]+\\) of file:\\(\\(/[a-z]:\\)?[^:]+\\):$" 2 1))
  "Alist of regexps to recognize error messages from `xmlrun'.
See `compilation-error-regexp-alist'."
  :type '(repeat :tag "Regexps"
		 (list (regexp :tag "Regexp")
		       (integer :tag "File index")
		       (integer :tag "Line index")
		       (choice :tag "Column index"
			       (integer)
			       (const :tag "None" nil))))
  :group 'xmlrun)

(defvar xmlrun-xml-source nil
  "*If non-nil, this is the name of the XML source file.")
(put 'xmlrun-xml-source 'xmlrun-type 'string)

(defvar xmlrun-result nil
  "*If non-nil, this is the name of the XSL result file.")
(put 'xmlrun-result 'xmlrun-type 'string)

(defvar xmlrun-command-history nil
  "The minibuffer history list for `xmlrun''s COMMAND argument.")
;;(make-variable-buffer-local 'xmlrun-command-history)

(eval-and-compile
  (autoload 'compile-internal "compile" ""))

(defun xmlrun-subst-expand-char (c parts)
  (cdr-safe (assq (downcase c) parts)))

(defun xmlrun-subst-expand (s parts)
  (loop for i from 0 to (1- (length s))
	as c = (aref s i)
	concat (if (eq c ?%)
		   (or (xmlrun-subst-expand-char (aref s (incf i)) parts)
		       (return nil)) 
		 (char-to-string (aref s i)))))

(defun xmlrun-populate-process-command-history ()
  (cond
   ((consp xmlrun-command)
    (let ((process-subst
	   (list
	    (cons ?b (and (buffer-file-name)
			  (file-name-nondirectory (buffer-file-name))))
	    (cons ?i xmlrun-input-file)
	    (cons ?s xmlrun-stylesheet-file)
	    (cons ?o xmlrun-output-file))))
      (setq xmlrun-command-history
	    (append
	     (mapcar (lambda (x)
		       (xmlrun-subst-expand x process-subst))
		     xmlrun-command)
	     xmlrun-command-history))))
   (t
    (apply 'format xmlrun-command
	   (if xmlrun-files
	       (funcall xmlrun-files)
	     (list (or xmlrun-xml-source "")
		   (let ((name (buffer-file-name)))
		     (if name
			 (file-name-nondirectory name)
		       ""))
		   (or xmlrun-result "")))))))

(defvar xmlrun-input-file nil
  "Filename of input file for `xmlrun' command")

(defvar xmlrun-input-history nil
  "The minibuffer history list for `xmlrun-get-process-parameters''s INPUT argument.")

(defvar xmlrun-stylesheet-file nil
  "Filename of stylesheet file for `xmlrun' command")

(defvar xmlrun-stylesheet-history nil
  "The minibuffer history list for `xmlrun-get-process-parameters''s STYLESHEET argument.")

(defvar xmlrun-output-file nil
  "Filename of output file for `xmlrun' command")

(defvar xmlrun-output-history nil
  "The minibuffer history list for `xmlrun-get-process-parameters''s OUTPUT argument.")

(defun xmlrun-get-process-parameters ()
  "Get and set the parameters for the `xmlrun' command"
  (interactive)
  (setq xmlrun-input-file
	(xmlrun-read-from-minibuffer "Input file: "
			      (file-name-nondirectory
			       (buffer-file-name))
			      'xmlrun-input-history))
  (setq xmlrun-stylesheet-file
	(xmlrun-read-from-minibuffer "Stylesheet file: "
			      (concat (file-name-sans-extension
				       (file-name-nondirectory
					(buffer-file-name)))
				      ".xsl")
			      'xmlrun-stylesheet-history))
  (setq xmlrun-output-file
	(xmlrun-read-from-minibuffer "Output file: "
			      (concat (file-name-sans-extension
				       (file-name-nondirectory
					xmlrun-input-file))
				      ".html")
			      'xmlrun-output-history)))

(defun xmlrun (command)
  "Process an XML file 

Runs COMMAND, a shell command, in a separate process asynchronously
with output going to the buffer *XML Process*.  

The first time that the program is run and whenever you provide a
prefix argument, e.g. \\[universal-argument] \\[xmlrun], prompts
for input filename, stylesheet file, and output filename.  Those
values are used with the templates in `xmlrun-command' to
populate this command's command history with the command lines to run
several XSLT processors using those values.  Use M-p and M-n to step
through the predefined commands, edit a command if necessary, or enter
a new command line.  The next time that this command is run, the
previously executed command is used as the default."
  (interactive
   (list (progn
	   (if (or
		current-prefix-arg
		(null xmlrun-command-history))
	       (progn
		 (xmlrun-get-process-parameters)
		 (xmlrun-populate-process-command-history)))
	   (read-from-minibuffer "Process command: "
				     (car xmlrun-command-history)
				     nil nil
				     'xmlrun-command-history))))
  (if xmlrun-offer-save
      (save-some-buffers nil nil))
  (compile-internal command "No more errors" "XML Process"
		    nil
		    xmlrun-error-regexps))

(defun xmlrun-read-from-minibuffer (prompt default history)
  "Read from minibuffer with default and command history."
(let ((value nil))
  (if (string-equal
       ""
       (setq value
	     (read-from-minibuffer (if default
				       (format
					"%s(default `%s') "
					prompt default)
				     (format "%s" prompt))
				   nil nil nil
				   history)))
	     default
	     value)))


(provide 'xmlrun)


tei-emacs/windoze.el
;; Windows-specific loading for tei-emacs project
;; C Wittern/S Rahtz April 2001, May 2004
;; Revised: $Date: 2004/09/23 13:23:11 $


;; dired stuff to open files a la Windows from Howard Melman
(defun dired-execute-file (&optional arg)
  (interactive "P")
  (mapcar #'(lambda (file)
      (w32-shell-execute "open" (convert-standard-filename file)))
          (dired-get-marked-files nil arg)))

(defun dired-mouse-execute-file (event)
  "In dired, execute the file or goto directory name you click on."
  (interactive "e")
  (set-buffer (window-buffer (posn-window (event-end event))))
  (goto-char (posn-point (event-end event)))
  (if (file-directory-p (dired-get-filename))
      (dired-find-file)
    (dired-execute-file)))
(global-set-key [?\C-x mouse-2] 'dired-mouse-execute-file)

(defun hrm-dired-mode-hook ()
  "Hook run when entering dired-mode."
    (define-key dired-mode-map "X" 'dired-execute-file)
    (define-key dired-mode-map [M-down-mouse-1] 'dired-mouse-mark-file))

(add-hook 'dired-mode-hook 'hrm-dired-mode-hook)


;; swap mouse-2 and mouse-3
(setq w32-swap-mouse-buttons t) 
(global-unset-key (kbd "<mouse-2>"))

;; Use standard Windows dialog box to open files
(load "dlgopen") 
;; (global-set-key "\C-x\C-f" 'dlgopen-open-files)



(setq w32-bdf-filename-alist
      (w32-find-bdf-fonts (concat teischemadir "/ucsfonts")))

(create-fontset-from-fontset-spec
 "-*-Courier New-normal-r-*-*-15-*-*-*-c-*-fontset-courier,
 arabic-iso8859-6:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-6
 chinese-big5-1:-*-MingLiU-normal-r-*-*-15-*-*-*-c-*-big5-*,
 chinese-big5-2:-*-MingLiU-normal-r-*-*-15-*-*-*-c-*-big5-*
 chinese-gb2312:-*-NSimSun-normal-r-*-*-15-*-*-*-c-*-gb2312-*,
 cyrillic-iso8859-5:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-5,
 greek-iso8859-7:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-7,
 indian-is13194:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-indian-is13194,
 japanese-jisx0208-1978:-*-MS Gothic-normal-r-*-*-15-*-*-*-c-*-jisx0208-sjis,
 japanese-jisx0208:-*-MS Gothic-normal-r-*-*-15-*-*-*-c-*-jisx0208-sjis,
 japanese-jisx0212:-*-MS Gothic-normal-r-*-*-15-*-*-*-c-*-jisx0212-sjis,
 katakana-jisx0201:-*-MS Gothic-normal-r-*-*-15-*-*-*-c-*-jisx0208-sjis,
 korean-ksc5601:-*-Gungsuh-normal-r-*-*-15-*-*-*-c-*-ksc5601-*,
 latin-iso8859-2:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-2,
 latin-iso8859-3:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-3,
 latin-iso8859-4:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-4,
 latin-iso8859-9:-*-Courier New-normal-r-*-*-15-*-*-*-c-*-iso8859-9,
 latin-jisx0201:-*-MS Gothic-normal-r-*-*-15-*-*-*-c-*-jisx0208-sjis,"
 t 'noerror)

(setq w32-enable-synthesized-fonts t)

(message "done loading windoze")


tei-emacs/xmlchars.el
;;; xmlchars.el --- Insert Unicode characters.

;; $Id: xmlchars.el,v 1.1.1.1 2004/09/23 13:23:17 rahtz Exp $

;; Most of these are taken from XML named characters from the ISO 8859 entity sets.
;; Hence the name.

;; Copyright (C) 2003 Norman Walsh

;; Author: Norman Walsh <ndw@nwalsh.com>
;; Maintainer: Norman Walsh <ndw@nwalsh.com>
;; Created: 2003-09-29
;; Version: 1.0
;; CVS ID: $Id: xmlchars.el,v 1.1.1.1 2004/09/23 13:23:17 rahtz Exp $
;; Keywords: utf-8 unicode xml characters

;; Inspired by http://www.tbray.org/ongoing/When/200x/2003/09/27/UniEmacs

;; This file is NOT part of GNU emacs.

;; This is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; This software is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.

;;; Commentary

;; This is all quite possibly a bit crude. But it works. I have a few more
;; ideas that I may implement if time allows. Suggestions welcome.

;;; Usage

;; 1. Load this file.
;; 2. Bind 'smart-double-quote, 'smart-single-quote, 'insert-xml-char,
;;    and 'shortcut-xml-char as you see fit.
;;
;;    I bind the quotes to " and ' in nxml-mode:
;;
;;      (define-key nxml-mode-map "\"" 'smart-double-quote)
;;      (define-key nxml-mode-map "\'" 'smart-single-quote)))
;;
;;    I bind the inserts to keys in my personal C-t map:
;;
;;      (define-key ctl-t-map "c" 'insert-special-char)
;;      (define-key ctl-t-map "e" 'insert-accented-char)))

;;; Changes

;; v1.1
;;   Added straight double and single quotes to the rotation
;; v1.0
;;   First release

;;; Code:

(defconst emacs-ldquo #x5397c)
(defconst emacs-rdquo #x5397d)
(defconst emacs-lsquo #x53978)
(defconst emacs-rsquo #x53979)
(defconst emacs-quot  #x0022)
(defconst emacs-apos  #x0027)

(defvar xml-chars-menu-alist
  '(("[nop]"     . 0)
    ("angst"     . #x212B)
    ("cent"      . #x00A2)
    ("copy"      . #x00A9)
    ("Dagger"    . #x2021)
    ("dagger"    . #x2020)
    ("deg"       . #x00B0)
    ("emsp"      . #x2003)
    ("ensp"      . #x2002)
    ("ETH"       . #x00D0)
    ("eth"       . #x00F0)
    ("euro"      . #x20AC)
    ("half"      . #x00BD)
    ("laquo"     . #x00AB)
    ("ldquo"     . #x201c)
    ("lsquo"     . #x2018)
    ("mdash"     . #x2014)
    ("micro"     . #x00B5)
    ("middot"    . #x00B7)
    ("ndash"     . #x2013)
    ("not"       . #x00AC)
    ("numsp"     . #x2007)
    ("para"      . #x00B6)
    ("permil"    . #x2030)
    ("puncsp"    . #x2008)
    ("raquo"     . #x00BB)
    ("rdquo"     . #x201d)
    ("rsquo"     . #x2019)
    ("reg"       . #x00AE)
    ("sect"      . #x00A7)
    ("THORN"     . #x00DE)
    ("thorn"     . #x00FE)
    ("trade"     . #x2122)
    ))

(defvar xml-chars-name-alist
  '(("AElig"     . #x00C6)
    ("Aacute"    . #x00C1)
    ("Acirc"     . #x00C2)
    ("Agrave"    . #x00C0)
    ("Aring"     . #x00C5)
    ("Atilde"    . #x00C3)
    ("Auml"      . #x00C4)
    ("Ccedil"    . #x00C7)
    ("Eacute"    . #x00C9)
    ("Ecirc"     . #x00CA)
    ("Egrave"    . #x00C8)
    ("Euml"      . #x00CB)
    ("Iacute"    . #x00CD)
    ("Icirc"     . #x00CE)
    ("Igrave"    . #x00CC)
    ("Iuml"      . #x00CF)
    ("Ntilde"    . #x00D1)
    ("Oacute"    . #x00D3)
    ("Ocirc"     . #x00D4)
    ("Ograve"    . #x00D2)
    ("Oslash"    . #x00D8)
    ("Otilde"    . #x00D5)
    ("Ouml"      . #x00D6)
    ("Uacute"    . #x00DA)
    ("Ucirc"     . #x00DB)
    ("Ugrave"    . #x00D9)
    ("Uuml"      . #x00DC)
    ("Yacute"    . #x00DD)
    ("aacute"    . #x00E1)
    ("acirc"     . #x00E2)
    ("aelig"     . #x00E6)
    ("agrave"    . #x00E0)
    ("aring"     . #x00E5)
    ("atilde"    . #x00E3)
    ("auml"      . #x00E4)
    ("ccedil"    . #x00E7)
    ("eacute"    . #x00E9)
    ("ecirc"     . #x00EA)
    ("egrave"    . #x00E8)
    ("euml"      . #x00EB)
    ("iacute"    . #x00ED)
    ("icirc"     . #x00EE)
    ("igrave"    . #x00EC)
    ("iuml"      . #x00EF)
    ("ntilde"    . #x00F1)
    ("oacute"    . #x00F3)
    ("ocirc"     . #x00F4)
    ("ograve"    . #x00F2)
    ("omacr"     . #x14d)
    ("oslash"    . #x00F8)
    ("otilde"    . #x00F5)
    ("ouml"      . #x00F6)
    ("szlig"     . #x00DF)
    ("uacute"    . #x00FA)
    ("ucirc"     . #x00FB)
    ("ugrave"    . #x00F9)
    ("uuml"      . #x00FC)
    ("yacute"    . #x00FD)
    ("yuml"      . #x00FF)
    ("frac12"    . #x00BD)
    ("frac13"    . #x2153)
    ("frac14"    . #x00BC)
    ("frac34"    . #x00BE)
    ("hellip"    . #x2026)
    ("iexcl"     . #x00A1)
    ("iquest"    . #x00BF)
    ("nbsp"      . #x00A0)
    ("plusmn"    . #x00B1)
    ("pound"     . #x00A3)
    ("yen"       . #x00A5)
    ))

(defvar xml-chars-shortcut-alist
  '(("AE"  . "AElig")
    ("A'"  . "Aacute")
    ("A^"  . "Acirc")
    ("A`"  . "Agrave")
    ("Ao"  . "Aring")
    ("A~"  . "Atilde")
    ("A\"" . "Auml")
    ("C,"  . "Ccedil")
    ("E'"  . "Eacute")
    ("E^"  . "Ecirc")
    ("E`"  . "Egrave")
    ("E\"" . "Euml")
    ("I'"  . "Iacute")
    ("I^"  . "Icirc")
    ("I`"  . "Igrave")
    ("I\"" . "Iuml")
    ("N~"  . "Ntilde")
    ("O'"  . "Oacute")
    ("O^" .  "Ocirc")
    ("O`"  . "Ograve")
    ("O/"  . "Oslash")
    ("O~"  . "Otilde")
    ("O\"" . "Ouml")
    ("U'"  . "Uacute")
    ("U^"  . "Ucirc")
    ("U`"  . "Ugrave")
    ("U\"" . "Uuml")
    ("Y'"  . "Yacute")
    ("a'"  . "aacute")
    ("a^"  . "acirc")
    ("ae"  . "aelig")
    ("a`"  .  "agrave")
    ("ao"  . "aring")
    ("a~"  . "atilde")
    ("a\"" . "auml")
    ("c,"  . "ccedil")
    ("e'"  . "eacute")
    ("e^"  . "ecirc")
    ("e`"  . "egrave")
    ("e\"" . "euml")
    ("i'"  . "iacute")
    ("i^"  . "icirc")
    ("i`"  . "igrave")
    ("i\"" . "iuml")
    ("n~"  . "ntilde")
    ("o'"  . "oacute")
    ("o^"  . "ocirc")
    ("o`"  . "ograve")
    ("o-"  . "omacr")
    ("o/"  . "oslash")
    ("o~"  . "otilde")
    ("o\"" . "ouml")
    ("sz"  . "szlig")
    ("u'"  . "uacute")
    ("u^"  . "ucirc")
    ("u`"  . "ugrave")
    ("u\"" . "uuml")
    ("y'"  . "yacute")
    ("y\"" . "yuml")
    ("12" . "frac12")
    ("13" . "frac13")
    ("14" . "frac14")
    ("15" . "frac15")
    ("16" . "frac16")
    ("18" . "frac18")
    ("23" . "frac23")
    ("25" . "frac25")
    ("34" . "frac34")
    ("35" . "frac35")
    ("38" . "frac38")
    ("45" . "frac45")
    ("56" . "frac56")
    ("58" . "frac58")
    ("78" . "frac78")
    ("<<" . "laquo")
    (".." . "hellip")
    ("!i" . "iexcl")
    ("?i" . "iquest")
    ("  " . "nbsp")
    ("+-" . "plusmn")
    ("$c" . "cent")
    ("$e" . "euro")
    ("$p" . "pound")
    ("$y" . "yen")
    ))

(defun  insert-xml-char (&optional charname)
  "Insert a special character by name; use a popup-menu if no name is given."
  (interactive)
  (let* ((xml-chars-menu
	  (list "Special chars" (append (list "") xml-chars-menu-alist)))
	 (value
	  (cond
	   ((and charname (assoc charname xml-chars-menu-alist))
	    (cdr (assoc charname xml-chars-menu-alist)))
	   ((and charname (assoc charname xml-chars-name-alist))
	    (cdr (assoc charname xml-chars-name-alist)))
	   (t
	    (x-popup-menu t
	     xml-chars-menu)))))
    (cond
     ((integerp value)
      (if (= 0 value)
	  (message "Cancelled")
	(ucs-insert value)))
     ((stringp  value) (insert value))
     (t (insert charname)))))

(defun shortcut-xml-char ()
  "Read a (two-character) keyboard shortcut and insert the corresponding character."
  (interactive)
  (let* ((c1 (read-char))
	 (c2 (read-char))
	 (str (concat (char-to-string c1) (char-to-string c2))))
    (cond
     ((assoc str xml-chars-shortcut-alist)
      (insert-xml-char (cdr (assoc str xml-chars-shortcut-alist))))
     (t (insert str)))))

(defun in-start-tag ()
  "Crude test to see if point is in side an open start tag."
  (interactive)
  (let (slim here pgt plt)
    (setq here (point))
    (setq slim
	  (if (> here 1024)
	      (- here 1024)
	    0))
    (setq pgt (search-backward ">" slim t))
    (goto-char here)
    (setq plt (search-backward "<" slim t))
    (goto-char here)
    (if (and pgt plt)
	(> plt pgt)
      plt)))

(defun smart-double-quote ()
  "Insert a left or right double quote as appropriate. Left quotes are inserted after a space, newline, or >. Right quotes are inserted after any other character, except if the preceding character is a quote, cycles through the three quote styles."
  (interactive)
  (let ((ch (char-before)))
    (cond
     ((in-start-tag)
      (insert "\""))
     ((or (not ch)
	  (char-equal ch ?>)
	  (char-equal ch 32)
	  (char-equal ch 10))
      (insert-xml-char "ldquo"))
     ((char-equal ch emacs-ldquo)
      (progn
	(delete-backward-char 1)
	(insert "\"")))
     ((char-equal ch emacs-quot)
      (progn
	(delete-backward-char 1)
	(insert-xml-char "rdquo")))
     ((char-equal ch emacs-rdquo)
      (progn
	(delete-backward-char 1)
	(insert-xml-char "ldquo")))
     ((char-equal ch emacs-ldquo)
      (progn
	(delete-backward-char 1)
	(insert-xml-char "rdquo")))
     ((char-equal ch emacs-lsquo)
      (insert-xml-char "ldquo"))
     (t (insert-xml-char "rdquo")))))

(defun smart-single-quote ()
  "Insert a left or right single quote as appropriate. Left quotes are inserted after a space, newline, or >. Right quotes are inserted after any other character, except if the preceding character is a quote, cycles through the three quote styles."
  (interactive)
  (let ((ch (char-before)))
    (cond
     ((in-start-tag)
      (insert "'"))
     ((or (not ch)
	  (char-equal ch ?>)
	  (char-equal ch 32)
	  (char-equal ch 10))
      (insert-xml-char "lsquo"))
     ((char-equal ch emacs-lsquo)
      (progn
	(delete-backward-char 1)
	(insert "'")))
     ((char-equal ch emacs-apos)
      (progn
	(delete-backward-char 1)
	(insert-xml-char "rsquo")))
     ((char-equal ch emacs-rsquo)
      (progn
	(delete-backward-char 1)
	(insert-xml-char "lsquo")))
     ((char-equal ch emacs-ldquo)
      (insert-xml-char "lsquo"))
     (t (insert-xml-char "rsquo")))))

;;; EOF
